File: persistence-jdbc/scaladsl/src/main/scala/com/lightbend/lagom/scaladsl/persistence/jdbc/JdbcSession.scala
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.scaladsl.persistence.jdbc

import java.sql.Connection

import scala.concurrent.Future

object JdbcSession {
  /**
   * Closes this resource after the block has finished.
   */
  def tryWith[Resource <: AutoCloseable, Out](resource: Resource)(block: Resource => Out): Out = {
    try {
      block(resource)
    } finally {
      resource.close()
    }
  }
}

trait JdbcSession {
  /**
   * Execute the given function with a connection.
   *
   * This will execute the callback in a thread pool that is specifically designed for use with JDBC calls.
   *
   * @param block The block to execute.
   * @return A future of the result.
   */
  def withConnection[T](block: Connection => T): Future[T]

  /**
   * Execute the given function in a transaction.
   *
   * This will execute the callback in a thread pool that is specifically designed for use with JDBC calls.
   *
   * @param block The block to execute.
   * @return A future of the result.
   */
  def withTransaction[T](block: Connection => T): Future[T]
}

File: persistence-jdbc/javadsl/src/main/java/com/lightbend/lagom/javadsl/persistence/jdbc/JdbcSession.java
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.javadsl.persistence.jdbc;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.concurrent.CompletionStage;

public interface JdbcSession {

    /**
     * Execute the given function with a connection.
     *
     * This will execute the callback in a thread pool that is specifically designed for use with JDBC calls.
     *
     * @param block The block to execute.
     * @return A future of the result.
     */
    <T> CompletionStage<T> withConnection(ConnectionFunction<T> block);

    /**
     * Execute the given function in a transaction.
     *
     * This will execute the callback in a thread pool that is specifically designed for use with JDBC calls.
     *
     * @param block The block to execute.
     * @return A future of the result.
     */
    <T> CompletionStage<T> withTransaction(ConnectionFunction<T> block);

    /**
     * SAM for using a connection.
     */
    @FunctionalInterface
    interface ConnectionFunction<T> {

        /**
         * Execute this function with the connection.
         *
         * @param connection The connection
         */
        T apply(Connection connection) throws SQLException;
    }
}
